tools/python/openlr/quality.py

```diff
@@ -0,0 +1,143 @@
+    λ1, λ2 = map(radians, [x.lat, y.lat])
+    Δφ = φ2 - φ1
+    Δλ = λ2 - λ1
+    a = sin(Δφ/2)**2 + cos(φ1) * cos(φ2) * sin(Δλ/2)**2
+    return 2 * atan2(a**0.5, (1 - a)**0.5)
+
+def polyline_length(ps):
+    if len(ps) < 2:
+        return 0.0
+    return sum(
+        distance(ps[i], ps[i - 1])
+        for i in range(1, len(ps))
+    )
+
+def common_subpath(l1, l2, eq=operator.eq):
+    prefs_len = [
+        [0 for _ in range(len(l2) + 1)]
+        for _ in range(len(l1) + 1)
+    ]
+    for i in range(1, len(l1) + 1):
+        for j in range(1, len(l2) + 1):
+            if eq(l1[i - 1], l2[j - 1]):
+                prefs_len[i][j] = prefs_len[i - 1][j - 1] + 1
+    l, i, j = max((
+        (prefs_len[i][j], i, j)
+        for i in range(1, len(l1) + 1)
+        for j in range(1, len(l2) + 1)
+    ))
+    return i - l, j - l, l
+
+def common_part(l1, l2):
+    s1, s2, l = common_subpath(l1, l2)
+    intersection = polyline_length(l1[s1:s1 + l + 1])
+    difference = (
+        polyline_length(l1[:s1 + 1]) +
+        polyline_length(l2[:s2 + 1]) +
+        polyline_length(l1[s1 + l - 1:]) +
+        polyline_length(l2[s2 + l - 1:])
+    )
+    total_length = intersection + difference
+    return 1.0 - intersection / total_length
+
+class Segment:
+    def __init__(self, segment_id, matched_route, golden_route):
+        #TODO(mgsergio): Remove this when deal with auto golden routes.
+        assert len(matched_route) > 1
+        assert len(golden_route) > 1
+        self.segment_id = segment_id
+        self.matched_route = matched_route
+        self.golden_route = golden_route or None
+
+    def __repr__(self):
+        return 'Segment({})'.format(self.segment_id)
+
+    def as_tuple(self):
+        return self.segment_id, self.matched_route, self.golden_route
+
+def parse_route(route):
+    if not route:
+        return None
+    result = []
+    for edge in route.findall('RoadEdge'):
+        start = edge.find('StartJunction')
+        result.append(
+            LatLon(float(start.find('lat').text), float(start.find('lon').text))
+        )
+    end = edge.find('EndJunction')
+    result.append(LatLon(float(end.find('lat').text), float(end.find('lon').text)))
+    return result
+
+def parse_segments(tree, limit):
+    segments = islice(tree.findall('.//Segment'), limit)
+    for s in segments:
+        ignored = s.find('Ignored')
+        if ignored is not None and ignored.text == 'true':
+            continue
+        segment_id = int(s.find('.//ReportSegmentID').text)
+        matched_route = parse_route(s.find('Route'))
+        # TODO(mgsergio): This is a temproraty hack. All untouched segments
+        # within limit are considered accurate, so golden path should be equal
+        # matched path.
+        golden_route = parse_route(s.find('GoldenRoute')) or matched_route
+        if not matched_route and not golden_route:
+            raise ValueError('At least one of golden route or mathched route should present')
```

Comment:
s/mathched/matched/
s/present/be present/

Reply:
Done.